File: apps/api/prachar_api/routers/unified_consult.py
"""Unified Consult Router — ONE set of endpoints for ALL domains.

Replaces the duplicated `/consult` (business) and `/creator/*` (creator) routers
with a single domain-agnostic router. The domain is selected by the `domain`
parameter, which loads the appropriate Domain Pack from the registry.

Endpoints:
  POST /consult              — universal consult (any domain)
  POST /consult/campaign     — universal campaign generation (any domain)
  POST /consult/tool/{tool_id} — domain-specific tool (e.g. repurpose, youtube_plan)
  GET  /consult/domains      — list available domains + subtypes (for onboarding UI)

The old /consult and /creator routers remain registered for backward compatibility
but delegate to this unified engine. Future domains (Restaurant, Clinic, etc.)
use ONLY this router — no new routers are created.
"""
from __future__ import annotations

import uuid
from typing import Any
from pydantic import BaseModel, Field

from fastapi import APIRouter, Depends, HTTPException, status

from prachar_shared.ai_gateway import BudgetExceeded
from prachar_shared.domain_packs import get_registry, register_all

from ..deps import CurrentUser, SessionDep
from ..infrastructure.consult_engine import ConsultEngine

# Ensure all packs are registered at import time
register_all()

router = APIRouter(prefix="/consult", tags=["consult"])


# ─── Request/Response schemas (domain-agnostic) ───────────────────────────


class ConsultRequest(BaseModel):
    """Universal consult request. Works for any domain."""
    message: str = Field(..., min_length=5, max_length=2000)
    domain: str = Field("business", description="Domain pack id: business, creator, restaurant, clinic")
    subtype_id: str = Field("", description="Optional subtype within the domain")
    brand_id: uuid.UUID | None = None


class CampaignRequest(BaseModel):
    """Universal campaign request. Works for any domain."""
    brand_id: uuid.UUID
    goal: str
    budget: str
    domain: str = Field("business", description="Domain pack id")


class ToolRequest(BaseModel):
    """Universal tool request. The tool is looked up in the domain's pack."""
    domain: str = Field(..., description="Domain pack id")
    inputs: dict[str, Any] = Field(default_factory=dict)


class DomainSummary(BaseModel):
    id: str
    label: str
    emoji: str
    customer_type: str
    subtypes: list[dict[str, Any]]


class DomainsResponse(BaseModel):
    domains: list[DomainSummary]


# ─── Endpoints ─────────────────────────────────────────────────────────────


@router.get("/domains", response_model=DomainsResponse, name="unified_consult_domains")
async def list_domains() -> DomainsResponse:
    """List all available domains and their subtypes.

    Used by the onboarding UI to render the type-selection screen.
    Adding a new domain pack automatically makes it appear here — no UI changes.
    """
    reg = get_registry()
    return DomainsResponse(
        domains=[
            DomainSummary(
                id=p.id,
                label=p.label,
                emoji=p.emoji,
                customer_type=p.customer_type,
                subtypes=[
                    {"id": s.id, "label": s.label, "emoji": s.emoji, "blurb": s.blurb}
                    for s in p.subtypes
                ],
            )
            for p in reg.all()
        ]
    )


@router.post("", name="unified_consult")
async def consult(
    body: ConsultRequest,
    user: CurrentUser,
    session: SessionDep,
) -> dict[str, Any]:
    """Universal consult endpoint. Works for ANY domain.

    Takes a free-text description + domain id and returns:
    - A conversational reply from PRACHAR AI
    - Domain-specific understanding (business / creator profile / etc.)
    - Growth opportunities
    - A 30-day plan
    - Extracted entity info
    - Brand id + name

    The pipeline is identical for all domains. Only the Domain Pack changes.
    """
    engine = ConsultEngine()
    try:
        result = await engine.consult(
            message=body.message,
            pack_id=body.domain,
            subtype_id=body.subtype_id,
            user=user,
            session=session,
            brand_id=body.brand_id,
        )
    except BudgetExceeded:
        return {
            "reply": (
                "Hey! I've hit my AI usage limit for this month. "
                "Contact your admin to upgrade your plan, and we'll pick up where we left off."
            ),
        }
    return {
        "reply": result.reply,
        "understanding": result.understanding,
        "opportunities": result.opportunities,
        "plan": result.plan,
        "extracted": result.extracted,
        "brand_id": result.brand_id,
        "brand_name": result.brand_name,
        "confidence": result.confidence,
        "tokens_used": result.tokens_used,
        "model": result.model,
        "domain": result.domain,
    }


@router.post("/campaign", name="unified_consult_campaign")
async def campaign(
    body: CampaignRequest,
    user: CurrentUser,
    session: SessionDep,
) -> dict[str, Any]:
    """Universal campaign generation endpoint. Works for ANY domain.

    Uses CampaignBrain.generate_campaign() (always) + the domain pack's
    campaign template + prompt to generate a presentation-deck preview.
    Persists the campaign plan.
    """
    engine = ConsultEngine()
    try:
        result = await engine.campaign(
            pack_id=body.domain,
            brand_id=body.brand_id,
            goal=body.goal,
            budget=body.budget,
            user=user,
            session=session,
        )
    except BudgetExceeded:
        return {
            "reply": "Hey! I've hit my AI usage limit for this month. Contact your admin to upgrade.",
        }
    return {
        "reply": result.reply,
        "preview": result.preview,
        "campaign_plan_id": result.campaign_plan_id,
        "confidence": result.confidence,
        "tokens_used": result.tokens_used,
        "model": result.model,
        "domain": result.domain,
    }


@router.post("/tool/{tool_id}", name="unified_consult_tool")
async def tool(
    tool_id: str,
    body: ToolRequest,
    user: CurrentUser,
) -> dict[str, Any]:
    """Invoke a domain-specific tool (e.g. repurpose, youtube_plan).

    The tool is looked up in the domain's pack. The pack supplies the prompt
    template; the engine handles the LLM call, JSON extraction, and response.

    This replaces the creator-specific /creator/repurpose and /creator/youtube-plan
    endpoints. New domains can add tools by defining ToolSpecs in their pack —
    no new endpoints needed.
    """
    engine = ConsultEngine()
    try:
        result = await engine.tool(
            pack_id=body.domain,
            tool_id=tool_id,
            inputs=body.inputs,
            user=user,
        )
    except BudgetExceeded:
        return {
            "reply": "Hey! I've hit my AI usage limit for this month.",
        }
    return {
        "reply": result.reply,
        "output": result.output,
        "tokens_used": result.tokens_used,
        "model": result.model,
        "tool_id": result.tool_id,
    }


@router.get("/nav/{domain}", name="unified_consult_nav")
async def get_nav(domain: str) -> dict[str, Any]:
    """Get the sidebar navigation for a domain.

    Used by the frontend to render the sidebar. Adding a new domain pack
    automatically makes its nav available here — no frontend hard-coding.
    """
    reg = get_registry()
    pack = reg.get(domain)
    if pack is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, f"Unknown domain: {domain}")
    return {
        "domain": pack.id,
        "label": pack.label,
        "emoji": pack.emoji,
        "nav_sections": [
            {
                "section": s.section,
                "items": [{"label": i.label, "path": i.path, "icon": i.icon} for i in s.items],
            }
            for s in pack.nav_sections
        ],
        "kpi_cards": [
            {"key": k.key, "label": k.label, "icon": k.icon, "hint": k.hint}
            for k in pack.kpi_cards
        ],
        "dashboard_widgets": [
            {"kind": w.kind, "title": w.title, "props": w.props}
            for w in pack.dashboard_widgets
        ],
        "quick_actions": [
            {
                "title": a.title,
                "description": a.description,
                "href": a.href,
                "icon": a.icon,
                "accent": a.accent,
            }
            for a in pack.quick_actions
        ],
        "tools": [
            {
                "id": t.id,
                "label": t.label,
                "description": t.description,
                "input_schema": t.input_schema,
            }
            for t in pack.tools
        ],
    }

File: apps/web-v2/src/lib/unified-consult.ts
/**
 * Unified Consult API client — ONE client for ALL domains.
 *
 * Replaces the duplicated `consultApi` and `creatorApi` clients. The domain is
 * selected by the `domain` parameter, which loads the appropriate Domain Pack
 * on the backend.
 *
 * Adding a new domain requires ZERO changes here. The backend registry handles it.
 */
import { apiGet, apiPost } from "./api";

// ─── Types (domain-agnostic) ──────────────────────────────────────────────

export interface DomainSubtype {
  id: string;
  label: string;
  emoji: string;
  blurb: string;
}

export interface DomainSummary {
  id: string;
  label: string;
  emoji: string;
  customer_type: "business" | "creator";
  subtypes: DomainSubtype[];
}

export interface NavItem {
  label: string;
  path: string;
  icon: string;
}

export interface NavSection {
  section: string;
  items: NavItem[];
}

export interface KpiCardSpec {
  key: string;
  label: string;
  icon: string;
  hint: string;
  /** Optional trend indicator. When absent, a graceful "connect channels" fallback is shown. */
  trend_direction?: "up" | "down" | "flat" | "new";
  /** Percentage change (e.g. 18 for "+18%"). Only shown when trend_direction is present. */
  trend_pct?: number;
  /** One-line context explaining what the number means and why it matters. */
  context?: string;
  /** Link href for a "See why" call-to-action. Shown only when the card is actionable. */
  see_why_href?: string;
}

export interface WidgetSpec {
  kind: string;
  title: string;
  props: Record<string, unknown>;
}

export interface ActionCardSpec {
  title: string;
  description: string;
  href: string;
  icon: string;
  accent: string;
}

export interface ToolSummary {
  id: string;
  label: string;
  description: string;
  input_schema: Record<string, unknown>;
}

export interface DomainConfig {
  domain: string;
  label: string;
  emoji: string;
  nav_sections: NavSection[];
  kpi_cards: KpiCardSpec[];
  dashboard_widgets: WidgetSpec[];
  quick_actions: ActionCardSpec[];
  tools: ToolSummary[];
}

export interface ConsultRequest {
  message: string;
  domain: string;
  subtype_id?: string;
  brand_id?: string;
}

export interface ConsultResponse {
  reply: string;
  understanding: Record<string, unknown>;
  opportunities: Array<Record<string, unknown>>;
  plan: Array<Record<string, unknown>>;
  extracted: Record<string, unknown>;
  brand_id: string;
  brand_name: string;
  confidence: number;
  tokens_used: number;
  model: string;
  domain: string;
}

export interface CampaignRequest {
  brand_id: string;
  goal: string;
  budget: string;
  domain: string;
}

export interface CampaignResponse {
  reply: string;
  preview: Record<string, unknown>;
  campaign_plan_id: string;
  confidence: number;
  tokens_used: number;
  model: string;
  domain: string;
}

// ─── Strategy types (B.1.1 + B.1.2) ───────────────────────────────────────

/** A single marketing strategy proposed by the StrategyEngine. */
export interface Strategy {
  /** Short, memorable name (e.g. "Signature Dish Hero"). */
  name: string;
  /** 2-3 sentences describing the strategic approach. */
  approach: string;
  /** 1-2 sentences explaining why this approach is effective. */
  why_it_works: string;
  /** 2-3 specific risks of this approach. */
  risks: string[];
  /** 1 sentence on the expected result if executed well. */
  expected_outcome: string;
  /** "primary" | "alternative" | "contrarian". */
  strategy_type: "primary" | "alternative" | "contrarian";
}

/** The "why A not B" explanation for the chosen primary strategy. */
export interface StrategyExplanation {
  /** The name of the chosen (primary) strategy. */
  chosen_strategy: string;
  /** 2-3 sentences explaining why this is the best choice. */
  reasoning: string;
  /** 1-2 sentences explaining why the alternative was not chosen. */
  why_not_alternative: string;
  /** 1-2 sentences explaining why the contrarian was not chosen. */
  why_not_contrarian: string;
  /** 3-5 specific factors that decided the choice. */
  key_factors: string[];
}

export interface ToolRequest {
  domain: string;
  inputs: Record<string, unknown>;
}

export interface ToolResponse {
  reply: string;
  output: Record<string, unknown>;
  tokens_used: number;
  model: string;
  tool_id: string;
}

// ─── API client ───────────────────────────────────────────────────────────

export const unifiedConsultApi = {
  /** List all available domains + subtypes (for onboarding UI). */
  async domains(): Promise<DomainSummary[]> {
    const res = await apiGet<{ domains: DomainSummary[] }>("/consult/domains");
    return res.domains;
  },

  /** Get the full config for a domain (nav, KPIs, widgets, tools). */
  async config(domain: string): Promise<DomainConfig> {
    return apiGet(`/consult/nav/${domain}`);
  },

  /** Universal consult — works for any domain. */
  async consult(req: ConsultRequest): Promise<ConsultResponse> {
    return apiPost("/consult", req);
  },

  /** Universal campaign generation — works for any domain. */
  async campaign(req: CampaignRequest): Promise<CampaignResponse> {
    return apiPost("/consult/campaign", req);
  },

  /** Invoke a domain-specific tool (e.g. repurpose, youtube_plan). */
  async tool(toolId: string, req: ToolRequest): Promise<ToolResponse> {
    return apiPost(`/consult/tool/${toolId}`, req);
  },
};
